File: examples/cs/request_reply/Request.cs
using System;
using Dragonfly;


namespace Request
{
    class Program
    {
        static void Main(string[] args)
        {
            bool KeepRunning = true;
            Module mod = new Module();

            string mm_ip = "localhost:7111";

            // Create a module and connect to MessageManager
            mod.ConnectToMMM(MID.REQUEST, mm_ip);

            // Subscribe to messages
            mod.Subscribe(MT.TEST_DATA);
            mod.Subscribe(MT.EXIT);

            Console.WriteLine("Request running...");

            Message m;
            while (KeepRunning)
            {
                try
                {
                    mod.SendSignal(MT.REQUEST_TEST_DATA);
                    Console.WriteLine("\nSent request for data");

                    m = mod.ReadMessage(Module.ReadType.Blocking);

                    if (m.msg_type == MT.TEST_DATA)
                    {
                        Console.WriteLine("Received message {0}", m.msg_type);

                        object o;
                        MDF.TEST_DATA data = new MDF.TEST_DATA();
                        o = data;
                        m.GetData(ref o);

                        Console.WriteLine("Data   a: {0},  b: {1},  x: {2} ", data.a, data.b, data.x);
                    }
                    else if (m.msg_type == MT.EXIT)
                    {
                        KeepRunning = false;
                    }
                }
                catch
                {
                    Console.WriteLine("ERROR: Something went wrong...");
                    KeepRunning = false;
                }

                System.Threading.Thread.Sleep(1000);
            }

            mod.DisconnectFromMMM();
        }
    }
}

File: examples/matlab/request_reply/Request.m
function Request(mm_ip)

    global DF;

    Dragonfly_BaseDir = '../../..';
    addpath([Dragonfly_BaseDir '/lang/matlab']);

    MessageTypes =  { ...
                        'TEST_DATA' ...
                        'EXIT' ...
                    };

    ConnectArgs = {'REQUEST', Dragonfly_BaseDir, ['./message_defs.mat']};
    if exist('mm_ip','var') && ~isempty(mm_ip)
        ConnectArgs{end+1} = ['-server_name ' mm_ip];
    end

    ConnectToMMM(ConnectArgs{:});
    Subscribe( MessageTypes{:})

    disp 'Request running..'


    while( 1)
        SendSignal REQUEST_TEST_DATA;
        fprintf('\nWaiting for reply\n');

        M = ReadMessage( 'blocking');
        switch(M.msg_type)
            case 'TEST_DATA'
                fprintf('Received  a=%d  b=%d  x=%f\n', M.data.a, M.data.b, M.data.x);
            
            case 'EXIT'
                break; 
        end
        
        pause(1);
    end

    DisconnectFromMMM

